### DISCUSSION TITLE
Add Dulmage-Mendelsohn decomposition to NetworkX

### DISCUSSION THREAD

**Discussion by Robbybp** [networkx/networkx#6720] [Algorithms]
Hi all,

I use the Dulmage-Mendelsohn decomposition frequently, and think it would make a nice addition to NetworkX. 

The Dulmage-Mendelsohn decomposition is a partition for bipartite graphs that gives all nodes that are unmatched in some maximum matching. It is very useful for analyzing systems of equations to determine over and under-constrained subsystems. 

I am thinking of a function with the following specification:
```python
def dulmage_mendelsohn_decomposition(G, top_nodes):
    """Return a partition of nodes in bipartite graph `G` into the subsets defined by the
    Dulmage-Mendelsohn decomposition

    Parameters
    ----------
    G : NetworkX graph
    top_nodes : container

    Returns
    -------
    top_partition : tuple

      Tuple of three elements: the top nodes that can possibly be unmatched, the
      top nodes that are matched in any maximum matching, and the top nodes matched
      with possibly unmatched bottom nodes.

    bot_partition : tuple

      Tuple of three elements: the bottom nodes matched with possibly unmatched top
      nodes, the bottom nodes that are matched in any maximum matching, and the bottom
      nodes that can be unmatched in some maximum matching

    """
    ...
```
I am thinking of adding this function to `networkx.algorithms.bipartite.matching`, as the algorithm could make use of the `_connected_by_alternating_paths` function in this module. My questions are:
1. Does this sound like a reasonable addition and location?
2. Any advice on the return types? My default for something like this would be two named tuples, but I am having trouble thinking of good names that apply to a general graph, as I usually think about this decomposition in terms of equations and variables, where "over-constrained", "under-constrained", and "well-constrained" make sense.

**Comment by dschult** [comment]
I'm having trouble understanding the output parameter description above because of phrases like: "the top nodes matched with possibly unmatched bottom nodes".  Perhaps the word "matched" is being used in two contexts here?

But a reference to the literature is a good place to start.  And an obvious reference from my perspective is the [wikipedia article on Dulmage-Mendelsohn decomposition](https://en.wikipedia.org/wiki/Dulmage%E2%80%93Mendelsohn_decomposition).  Can you use their names for the returned sets?  

It seems like inputs would need to be `G`, `top_nodes`, `matching` and I don't know if it should include the set of nodes not in the matching, or if that should be computed by the function.  Then I'm guessing the output would be the "even", "odd" and "unreachable" sets of the nodes. I don't know whether that is best returned as 3 sets, or as a dict keyed by node to an indicator string "even", "odd", or "unreachable". 

This is reasonable to include in NetworkX and the module you mention looks like the right place for it. :)

**Comment by Robbybp** [comment]
I've thought about it a little more, and I think the following is the simplest specification that provides all the information a user would need:
```python
def dulmage_mendelsohn_decomposition(G, top_nodes=top_nodes, matching=matching):
    """Return a partition of nodes in bipartite graph `G` into the subsets defined by the
    Dulmage-Mendelsohn decomposition

    Parameters
    ----------
    G : NetworkX graph
    top_nodes : container

    Returns
    -------
    reachable_from_top : list

        List of nodes reachable by an alternating path from unmatched top nodes

    reachable_from_bot : list

        List of nodes reachable by an alternating path from unmatched bottom
        nodes

    unreachable : list

        List of nodes that are not reachable by an alternating path from
        unmatched nodes.


    """
    ...
```
`top_nodes` and `matching` are not strictly necessary arguments, as we can always compute these. However, if choice of `top_nodes` is not provided, the selection of these is not unique. 

The algorithm partitions the nodes into three subsets, which are easier to explain than the six subsets I first proposed. However, it is now a bit more work to partition these subsets into the two bipartite sets, which we usually (always?) want to do when using this decomposition.

I'll open a PR with a draft implementation of this shortly.

**Reply by Robbybp** [reply]
I have implemented this in #6730. We can take any further discussion there.
